gr-blocks/python/blocks/qa_rotator_cc.py

```diff
@@ -0,0 +1,199 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+#
+# Copyright 2019 Free Software Foundation, Inc.
+#
+# This file is part of GNU Radio
+#
+# GNU Radio is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 3, or (at your option)
+# any later version.
+#
+# GNU Radio is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with GNU Radio; see the file COPYING.  If not, write to
+# the Free Software Foundation, Inc., 51 Franklin Street,
+# Boston, MA 02110-1301, USA.
+#
+
+from gnuradio import gr, gr_unittest
+from gnuradio import blocks
+import numpy as np
+import pmt
+import time
+
+
+class qa_rotator_cc(gr_unittest.TestCase):
+
+    def setUp(self):
+        self.n_samples  = 100
+        self.f_in       = 0.01  # frequency on rotator's input
+        self.f_shift    = 0.015 # rotator's starting frequency
+        tag_inc_updates = True
+
+        self.tb = gr.top_block()
+
+        # Input IQ samples
+        in_angles  = 2 * np.pi * np.arange(self.n_samples) * self.f_in
+        in_samples = np.exp(1j*in_angles)
+
+        # Rotator's starting phase increment
+        phase_inc = 2 * np.pi * self.f_shift
+
+        self.source     = blocks.vector_source_c(in_samples)
+        self.rotator_cc = blocks.rotator_cc(phase_inc, tag_inc_updates)
+        self.sink       = blocks.vector_sink_c()
+
+    def tearDown(self):
+        self.tb = None
+
+    def test_001_t(self):
+        """Complex sinusoid frequency shift
+        """
+
+        f_out = self.f_in + self.f_shift # frequency of rotator's output
+
+        # Expected IQ samples
+        expected_angles  = 2 * np.pi * np.arange(self.n_samples) * f_out
+        expected_samples = np.exp(1j*expected_angles)
+
+        self.tb.connect(self.source, self.rotator_cc, self.sink)
+        self.tb.run()
+
+        for (x,y) in zip(self.sink.data(), expected_samples):
+            self.assertComplexAlmostEqual(x, y, places=5)
```

Comment:
I am aware that there is `assertComplexTuplesAlmostEqual`. However, I couldn't get it to work. It would always pass even when it shouldn't. I couldn't understand why.

Reply:
Meh. The macro is supposed to save you time developing, and it's clearly not, so let's move on.